File: android/app/src/androidTest/java/extentions/PerformanceLogger.java
package extentions;

import android.util.Log;

import org.json.JSONException;
import org.json.JSONObject;

public class PerformanceLogger {
    private final String tag;
    private String method;
    private long startTime;
    private boolean started;
    private boolean ended;
    private JSONObject report;

    public PerformanceLogger(final String tagName) {
        tag = tagName;

        try {
            JSONObject reportObject = new JSONObject();
            reportObject.put("TAG", tag);
            report = reportObject;
        } catch (JSONException e) {
            Log.e(tag, "Performance logger Unable to init report object");
        }
    }

    public void start(final String methodName) {
        if (started && !ended) {
            throw new RuntimeException("Performance logger already started, stop first!");
        }

        started = true;
        ended = false;
        method = methodName;
        startTime = System.currentTimeMillis();
    }

    public void end(final String methodName) {
        try {
            if (!started) {
                throw new RuntimeException("Performance logger is not started!");
            }

            if (!method.equals(methodName)) {
                throw new RuntimeException("Performance logger tries to stop unknown method!");
            }

            started = false;
            ended = true;


            // add to the report
            report.put(method, System.currentTimeMillis() - startTime);
        } catch (JSONException e) {
            Log.e(tag, "Performance logger Unable to end the report");
        }
    }

    public void log() {
        Log.d(tag, report.toString());
    }
}

File: ios/XamanTests/PerformanceLogger.m
//
//  PerformanceLogger.m
//  XamanTests
//
//  Created by XRPL-Labs on 05/10/2022.
//

#import <Foundation/Foundation.h>

#import "PerformanceLogger.h"


@implementation PerformanceLogger

- (id)initWithTag:(NSString *)tagName {
  self = [super init];
  
  tag = tagName;
  
  report = [[NSMutableDictionary alloc] init];
  [report setValue:tagName forKey:@"TAG"];
  
  return self;
}


- (void)start:(NSString *)methodName{
  if(started && !ended){
    @throw [NSException exceptionWithName:@"Performance Logger"
                                   reason:@"Performance logger already started, stop first!"
                                 userInfo:nil];
  }
  
  started = YES;
  ended = NO;
  method = methodName;
  startTime = [NSDate date];
}

- (void)end:(NSString *)methodName{
  if(!started){
    @throw [NSException exceptionWithName:@"Performance Logger"
                                   reason:@"Performance logger is not started!"
                                 userInfo:nil];
  }
  
  if(![method isEqualToString:methodName]){
    @throw [NSException exceptionWithName:@"Performance Logger"
                                   reason:@"Performance logger tries to stop unknown method!"
                                 userInfo:nil];
  }
  
  started = FALSE;
  ended = TRUE;
  
  NSInteger executionTime = fmod([[NSDate date] timeIntervalSinceDate:startTime], 1) * 1000;
  
  [report setValue:[NSNumber numberWithDouble:executionTime] forKey:method];
}


- (void)log {
  NSError *error;
  NSData *reportData = [NSJSONSerialization  dataWithJSONObject:report options:0 error:&error];
  
  if(error != nil){
    @throw error;
  }
  
  NSString *reportString = [[NSString alloc] initWithData:reportData encoding:NSUTF8StringEncoding];
  
  NSLog(@"%@", reportString);
}



@end

